libp2p/muxers/mplex/mplex.nim

```diff
@@ -63,96 +67,128 @@ proc newStreamInternal*(m: Mplex,
+  except CancelledError as exc:
+    trace "cancling stream handler", exc = exc.msg
+    await chann.reset()
+    raise
   except CatchableError as exc:
     trace "exception in stream handler", exc = exc.msg
     await chann.reset()
+    await m.cleanupChann(chann)
 
 method handle*(m: Mplex) {.async, gcsafe.} =
-  trace "starting mplex main loop", oid = m.oid
+  trace "starting mplex main loop", oid = $m.oid
   try:
-    try:
-      while not m.connection.closed:
-        trace "waiting for data", oid = m.oid
-        let (id, msgType, data) = await m.connection.readMsg()
-        trace "read message from connection", id = id,
-                                              msgType = msgType,
-                                              data = data.shortLog,
-                                              oid = m.oid
-
-        let initiator = bool(ord(msgType) and 1)
-        var channel: LPChannel
-        if MessageType(msgType) != MessageType.New:
-          let channels = m.getChannelList(initiator)
-          if id notin channels:
-
-            trace "Channel not found, skipping", id = id,
-                                                 initiator = initiator,
-                                                 msg = msgType,
-                                                 oid = m.oid
-            continue
-          channel = channels[id]
-
-        logScope:
-          id = id
-          initiator = initiator
-          msgType = msgType
-          size = data.len
-          muxer_oid = m.oid
-
-        case msgType:
-          of MessageType.New:
-            let name = string.fromBytes(data)
-            channel = await m.newStreamInternal(false, id, name)
-
-            trace "created channel", name = channel.name,
-                                     oid = channel.oid
-
-            if not isNil(m.streamHandler):
-              # launch handler task
-              asyncCheck m.handleStream(channel)
-
-          of MessageType.MsgIn, MessageType.MsgOut:
-            logScope:
-              name = channel.name
-              oid = channel.oid
-
-            trace "pushing data to channel"
-
-            if data.len > MaxMsgSize:
-              raise newLPStreamLimitError()
-            await channel.pushTo(data)
-          of MessageType.CloseIn, MessageType.CloseOut:
-            logScope:
-              name = channel.name
-              oid = channel.oid
-
-            trace "closing channel"
-
-            await channel.closeRemote()
-            m.getChannelList(initiator).del(id)
-            trace "deleted channel"
-          of MessageType.ResetIn, MessageType.ResetOut:
-            logScope:
-              name = channel.name
-              oid = channel.oid
-
-            trace "resetting channel"
-
-            await channel.reset()
-            m.getChannelList(initiator).del(id)
-            trace "deleted channel"
-    finally:
-      trace "stopping mplex main loop", oid = m.oid
+    defer:
```

Comment:
`defer` inside of `try` block you are trying to break it really, `finally` here is much more natural from my point of view.

Reply:
With all things being equal, I think `defer` is cleaner than a bunch of nested `try/finally`, but it is a matter of preference. Hopefully, defer wont break this time tho 😄

Reply:
fwiw, `defer` has been proposed for removal from the language and has proven buggy before - it will not get as much attention on the compiler side.. `finally` might be the safer option overall